Add additional process attributes to registry

mjwolf opened the pull request:
## Changes

This PR adds additional fields from the [ECS process schema](https://github.com/elastic/ecs/blob/main/schemas/process.yml) to the process registry. The fields added are mainly based on the POSIX process model, and will allow sending enhanced process telemetry from these systems.

## Merge requirement checklist

* [x] [CONTRIBUTING.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md) guidelines followed.
* [x] [CHANGELOG.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CHANGELOG.md) updated for non-trivial changes.
* [x] [schema-next.yaml](https://github.com/open-telemetry/semantic-conventions/blob/main/schema-next.yaml) updated with changes to existing conventions.

trisch-me reviewed (approved):
thanks @mjwolf, LGTM

joaopgrassi commented:
FYI @open-telemetry/semconv-system-approvers

braydonk commented on docs/attributes-registry/process.md:
```diff
@@ -11,12 +11,32 @@
 | `process.command` | string | The command used to launch the process (i.e. the command name). On Linux based systems, can be set to the zeroth string in `proc/[pid]/cmdline`. On Windows, can be set to the first parameter extracted from `GetCommandLineW`. | `cmd/otelcol` |
 | `process.command_args` | string[] | All the command arguments (including the command/executable itself) as received by the process. On Linux-based systems (and some other Unixoid systems supporting procfs), can be set according to the list of null-delimited strings extracted from `proc/[pid]/cmdline`. For libc-based executables, this would be the full argv vector passed to `main`. | `[cmd/oteco ...
 | `process.command_line` | string | The full command used to launch the process as a single string representing the full command. On Windows, can be set to the result of `GetCommandLineW`. Do not set this if you have to assemble it just for monitoring; use `process.command_args` instead. | `C:\cmd\otecol --config="my directory\config.yaml"` |
+| `process.end` | string | The date and time the process ended, in ISO 8601 format. | `2023-11-21T09:26:12.315Z` |
```
I think this and `process.exit_code` are strange as resource attributes. 

If it's a long running process, there is no end or exit code yet. Once we have them, I guess they could be added as resource attributes, but then that means the resource has tangibly changed even though it's representing the same resource. So I don't see how these are useful as resource attributes, since resources are sort of expected to be immutable.

mjwolf replied:
These fields are only in the attributes-registry, and not in the process resource schema, for this reason. You're right that it wouldn't really make sense to have these values in the immutable resource when they probably wouldn't be known at creation time, and couldn't be added later because of resource immutability.

The intention of having these in the attributes-registry is they could be used, for example, as attributes on a "process-exit" event log, where the values are known.

AlexanderWert replied:
@mjwolf We would need to split the yaml files into two groups then (for resource and non-resource).

I think the best way to do so, is to move existing process attributes to registry, first. and then only extend the registry yaml files.

AlexanderWert replied:
The problem here is that the entire group (in yaml) is defined as resource type.
To fix that properly, we need to resolve https://github.com/open-telemetry/build-tools/issues/242 first. 
So, I fear this is another case being blocked on https://github.com/open-telemetry/build-tools/issues/242.

Which however should not block us on discussing the general things here.

trisch-me replied:
@braydonk I would also add that these are not resource attributes despite they are defined in the yaml file as such. There were just normal attributes and change to `resource` was a quick fix until a solution for building the tools will be implemented, see https://github.com/open-telemetry/semantic-conventions/pull/524

braydonk replied:
Thanks for the clarification, I didn't know about this issue with the build tools and mistook this to mean they were intended to be resource attributes. Assuming that they are not intended to be resource attributes and would be changed once that issue is resolved, then this is fine with me.

mjwolf replied:
https://github.com/open-telemetry/build-tools/issues/242 shouldn't be blocking this any longer

dmitryax commented on docs/resource/process.md:
```diff
@@ -32,10 +32,12 @@
 | [`process.executable.path`](../attributes-registry/process.md) | string | The full path to the process executable. On Linux based systems, can be set to the target of `proc/[pid]/exe`. On Windows, can be set to the result of `GetProcessImageFileNameW`. | `/usr/bin/cmd/otelcol` | Conditionally Required: See alternative attributes below. |
-| [`process.owner`](../attributes-registry/process.md) | string | The username of the user that owns the process. | `root` | Recommended |
-| [`process.parent_pid`](../attributes-registry/process.md) | int | Parent Process identifier (PID). | `111` | Recommended |
+| [`process.owner`](../attributes-registry/process.md) | string | The username of the user that owns the process. [1] | `root` | Recommended |
+| [`process.parent_pid`](../attributes-registry/process.md) | int | Parent Process identifier (PPID). | `111` | Recommended |
```
Why only these two got updated, and other new attributes not added to this file? Was this done manually?

Also, changes to the existing attributes are probably better to be done in a separate PR since it's not aligned with this PR's goal

dmitryax replied:
> I'll take out the note on process.owner, I agree that's probably worth its own discussion and PR, but I'd like to leave in PPID. PPID is the correct abbreviation for this, and I think it was probably a mistake that it was PID before.

Sure, we need to fix this. But it's still outside of this PR's goal. So can be another easy-to-merge bugfix PR.

mjwolf replied:
I've taken out both those changes

mjwolf commented on model/registry/process.yaml:
```diff
@@ -10,11 +10,32 @@ groups:
           Process identifier (PID).
         examples: [1234]
-      - id: parent_pid
+      - id: parent.pid
```
I've taken out the renaming of `parent_pid` from this PR

trisch-me commented on docs/attributes-registry/process.md:
```diff
@@ -11,12 +11,31 @@
+| `process.vpid` | int | Virtual process identifier. [2] | `12` |
+
+**[1]:** As environment variables may change during a process's lifespan, this should be captured as a snapshot when the event occurred.
+May be filtered to protect sensitive information.
```
I think it should be then `MAY BE` or even `MUST BE`. We need to avoid sensitive information being captured

trisch-me replied:
Schema is the right place to specify these things related to the privacy/sensitive information. See for example [url schema](https://github.com/open-telemetry/semantic-conventions/blob/main/model/registry/url.yaml#L20)
Also imperative statements should be written with upper case

braydonk replied:
I think I would vote for `SHOULD` in this case. Intuitively we know there are all kinds of things that *could* be considered sensitive information, but if we can't give a definitive list of the things that must be filtered it, then it feels weird to give it the `MUST` imperative since it's hard to be specific.

Arguably this could apply to URL as well, where the decision was to use `MUST`. So that decision could be used to inform this. In this case, sensitive information could be any manner of credentials, keys, PII, and they generally don't take any standard form. So it feels different than with URL where it seemed specific to username and password.

I don't feel too strongly, but I think I would choose `SHOULD` or `MAY` personally, unless we found a way to be more prescriptive here about exactly what type of sensitive info should be filtered out (rather than leaving it up to the implementation).

mjwolf replied:
I've changed this to `SHOULD` now

joaopgrassi replied:
Note that `SHOULD` actually means to avoid implementing it, there must be reasons. There was a recent discussion about the "weight" if SHOULD as it's commonly interpreted as "optional" and "not really required".  

>SHOULD   This word, or the adjective "RECOMMENDED", mean that there
   may exist valid reasons in particular circumstances to ignore a
   particular item, but the full implications must be understood and
   carefully weighed before choosing a different course.
   
https://datatracker.ietf.org/doc/html/rfc2119

mjwolf replied:
> Note that `SHOULD` actually means to avoid implementing it, there must be reasons. There was a recent discussion about the "weight" if SHOULD as it's commonly interpreted as "optional" and "not really required", which is wrong.

I feel like using `SHOULD` here matches this definition well. Given the security implications of unredacted env vars, users should understand the risks of not filtering, but still, there might be some use cases where the full data is wanted.

mjwolf replied:
I've taken `env_vars` out of this PR, and created https://github.com/open-telemetry/semantic-conventions/issues/672 to track adding it in the future, so this discussion no longer applies to this PR

joaopgrassi commented on docs/attributes-registry/process.md:
```diff
@@ -11,12 +11,32 @@
 | `process.command_args` | string[] | All the command arguments (including the command/executable itself) as received by the process. On Linux-based systems (and some other Unixoid systems supporting procfs), can be set according to the list of null-delimited strings extracted from `proc/[pid]/cmdline`. For libc-based executables, this would be the full argv vector passed to `main`. | `[cmd/oteco ...
 | `process.command_line` | string | The full command used to launch the process as a single string representing the full command. On Windows, can be set to the result of `GetCommandLineW`. Do not set this if you have to assemble it just for monitoring; use `process.command_args` instead. | `C:\cmd\otecol --config="my directory\config.yaml"` |
+| `process.end` | string | The date and time the process ended, in ISO 8601 format. | `2023-11-21T09:26:12.315Z` |
+| `process.env_vars` | string[] | Array of environment variable bindings. [1] | `[PATH=/usr/local/bin;/usr/bin, USER=ubuntu]` |
```
I see this is marked as `beta` in ECS. Does any of you know if it was intended to take all (even beta) fields from ECS? Are those in use by ECS customers/users? Without knowing what it is used today, it's kinda hard to see any value.. specially given SDKs may need to even do extra things to protect the value from it.

mjwolf replied:
This field is being used by Elastic users now. One example of a current use case is that there are security detections on the use of `LD_PRELOAD` and `LD_LIBRARY_PATH`.

I'm not sure right now if the ECS merger should take beta fields, I'll try to find out

mjwolf replied:
Hi @joaopgrassi, looking at the [OTEP to merge ECS](https://github.com/open-telemetry/oteps/blob/main/text/0199-support-elastic-common-schema-in-opentelemetry.md), it says `OTel SemConv will adopt ECS in its full scope (except for individual adjustments in detail where inevitable)`, so maybe beta fields can be taken, I'm not sure, it's a bit vague about what "full scope", or "individual adjustments" means. I couldn't find anything referencing beta fields specifically.

For this particular field, `process.env_vars` will be promoted from beta to GA in the next ECS release https://github.com/elastic/ecs/pull/2315

svrnm replied:
I just stumbled upon this while searching for a standardized way to store (specific) environment variables in resource attributes. To be honest I am surprised about the suggested format:

```
[PATH=/usr/local/bin;/usr/bin, USER=ubuntu]`
```

Similar to HTTP headers I would rather have expected a format like

```
| `process.environment_variable.<key>` | string | Process environment variables, `<key>` being the environment variable name, the value being the environment variable value. | `proccess.environment_variable.PATH="/usr/local/bin;/usr/bin"`; `process.environment_variable.USER="ubuntu"` | Opt-In |
```

This makes it much easier to (a) find environment variables in the backend (or collector) and (b) only collect those environment variables I really care about (and to create filters to remove those not wanted)

jsuereth replied:
+1 to this suggestion for consistency.

mjwolf replied:
Hi @svrnm, thanks for the suggestion. I think this will need more discussion, so I've taken `env_vars` out of this PR, and created an [issue](https://github.com/open-telemetry/semantic-conventions/issues/672) to decide how to implement this before it's added in the future

svrnm replied:
thanks @mjwolf , that totally makes sense to have this in a separate discussion, I will follow up there

mjwolf commented:
I've removed `env_vars` from this PR, which I think was the source all the open discussions on this PR, and created https://github.com/open-telemetry/semantic-conventions/issues/672 to track and discuss it in the future

mjwolf commented on CHANGELOG.md:
```diff
@@ -80,6 +80,8 @@ release.
   ([#21](https://github.com/open-telemetry/semantic-conventions/pull/21))
 - Add `messaging.gcp_pubsub.message.ordering_key` attribute.
   ([#528](https://github.com/open-telemetry/semantic-conventions/pull/528))
+- Add additional attributes to process attribute registry.
```
Thanks, I moved this back to unreleased. It must have move section by a merge, and I didn't notice

mjwolf commented on schema-next.yaml:
```diff
@@ -11,6 +11,10 @@ versions:
+    # https://github.com/open-telemetry/semantic-conventions/pull/564
+    - rename_attributes:
+        attribute_map:
+          process.parent_pid: process.parent.pid
```
This schema change has been removed

svrnm commented on model/registry/process.yaml:
```diff
@@ -76,3 +127,22 @@ groups:
           An additional description about the runtime of the process, for example
           a specific vendor customization of the runtime environment.
         examples: 'Eclipse OpenJ9 Eclipse OpenJ9 VM openj9-0.21.0'
+      - id: start
+        type: string
+        brief: >
+          The date and time the process started, in ISO 8601 format.
+        examples: ['2023-11-21T09:25:34.853Z']
+      - id: end
+        type: string
+        brief: >
+          The date and time the process ended, in ISO 8601 format.
+        examples: ['2023-11-21T09:26:12.315Z']
```
https://github.com/open-telemetry/semantic-conventions/issues/660

Linking this with this issue, since this will introduce another 2 instances of date time values

jsuereth replied:
I'd like to resolve date/time recommendations early here.  I.e. let's not continue digging a hole of inconsistent date/time representation.  

Ideally, I think all date/timestamps should be representing similarly to *raw* timestamps in OTLP.

svrnm replied:
What does raw timestamps mean in that context? Looking into OTLP docs/code I guess unix timestamp in nano seconds (in a 64bit int)?

jsuereth replied:
Yeah, I mean unix timestamp in nanoseconds.

I'd like to hold of merging this until we can resovle this dicussion.

svrnm replied:
@jsuereth should we move that conversation over to https://github.com/open-telemetry/semantic-conventions/issues/660 then or is there another issue prior to that where this has been discussed?

jsuereth replied:
Sure, #660 is a decent place. We had some in person discussions in the Specification meeting and the Semconv meeting.

I'm not sure we have consensus here, but I *do* think we can likely make progress by sticking (consistently) to String-dates in ISO 8601 format in semconv (what you have).

jsuereth replied:
Let's discuss monday to gain consensus.

I think we could take a stand that no matter what the outcome of #660 is, we'll ensure *on the wire* ISO-8601 strings is what people see from OTel semconv.  This would then just be a limitation we place on the solution space for #660.  From what I gathered in both discussions, I think this would be acceptable, but want to run it past the working-group.

trisch-me replied:
@jsuereth do we have an outcome that we are staying with ISO-8601 string and this conversation could be resolved? Or you want additional change in this PR regarding time format?

mjwolf replied:
Hi @jsuereth, has this been resolved? Should we stay with IS0-8601 in the registry schema?

jsuereth replied:
Wow, sorry for not updating this.  

YES sticking with ISO-8601 was the consensus.  Regardless of future work, we think this *is* the representation we want for times as it's the most user-friendly across the most backends.  Any 'advanced' types we may support will need to deal with the decision that, within OTLP, we want ISO-8601 strings.

Thought I had commented on this bug, but apparently just dropped the ball, much apologies!

mjwolf commented:
> @mjwolf could you please update the PR to use [new way ](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md#adding-a-changelog-entry) of adding a changelog entry?

I've changed to use the new changelog method

lmolkova commented on docs/attributes-registry/process.md:
```diff
@@ -11,12 +11,27 @@
 | `process.command` | string | The command used to launch the process (i.e. the command name). On Linux based systems, can be set to the zeroth string in `proc/[pid]/cmdline`. On Windows, can be set to the first parameter extracted from `GetCommandLineW`. | `cmd/otelcol` |
 | `process.command_args` | string[] | All the command arguments (including the command/executable itself) as received by the process. On Linux-based systems (and some other Unixoid systems supporting procfs), can be set according to the list of null-delimited strings extracted from `proc/[pid]/cmdline`. For libc-based executables, this would be the full argv vector passed to `main`. | `[cmd/oteco ...
 | `process.command_line` | string | The full command used to launch the process as a single string representing the full command. On Windows, can be set to the result of `GetCommandLineW`. Do not set this if you have to assemble it just for monitoring; use `process.command_args` instead. | `C:\cmd\otecol --config="my directory\config.yaml"` |
+| `process.end` | string | The date and time the process ended, in ISO 8601 format. | `2023-11-21T09:26:12.315Z` |
```
would it make sense to call it `exit` since it's already used in `exit_code` and more frequently used in relation to processes. Then we can make it a namespace:

`process.exit.time`
`process.exit.code`

We might have more over time
`process.exit.reason`
...

mjwolf replied:
Hi @lmolkova, what do you think of keeping `process.end` as is, and leaving the `user` field until the user namespace is available?

I'd like to get everything resolved and move this PR into the Needs Approvals state on the project board soon

lmolkova replied:
Exit is a common command that used to terminate the process ([linux](https://www.google.com/search?q=linux+process+exit&rlz=1C1RXQR_enUS1082US1082&oq=linux+process+exit&gs_lcrp=EgZjaHJvbWUyBggAEEUYOTIGCAEQIxgnMgcIAhAAGIAEMggIAxAAGBYYHjIICAQQABgWGB4yCAgFEAAYFhgeMggIBhAAGBYYHjIGCAcQRRg80gEINTQ5OGowajeoAgCwAgA&sourceid=chrome&ie=UTF-8), [windows](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-exitprocess)). 


If we designed these attributes from the scratch, we'd at least entertained the idea of calling them `process.start_time` and `exit_time` since they are too generic without `_time` and since `exit` is much more common term comparing to `end` in the context of processes.

So it seems having `start` and `exit` as namespaces and (`process.start.time`, `process.exit.time` and `process.exit.code`) would provide more clarify and allow to add start/stop info to specific namespaces.

trisch-me replied:
I should note that if we go this part, we will not be able to use `start` as a namespace because we already have this field as simple type in ECS and according to our [rules](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md#merging-existing-ecs-conventions) we shouldn't change the type (use existing ECS name as a namespace).
With `exit.*` vs `end` we don't have this problem.

In general I am not against changing it to the new structure, but we should be careful with naming

mjwolf replied:
I could change to `process.creation` for the namespace instead of start.  Creation is in the [Windows API](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessa) and Linux docs (with Linux, the syscalls are fork and exec, not creation specifically;  there are still some places that refer to process creation, like [this](https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v1/pids.html?highlight=process+creation))

@lmolkova, @trisch-me, if you agree, I'll change to namespaces for `creation` and `end`.

mjwolf replied:
> So there will be then `creation.time`, `exit.time` and `exit.code`?
> 
> I'm ok with this naming. Let's see what others think

Yes, that's what the fields will be renamed to

lmolkova replied:
Thank you, sounds good!

mjwolf replied:
I've changed the fields to `creation.time`, `exit.time` and `exit.code`

lmolkova commented on docs/attributes-registry/process.md:
```diff
@@ -11,12 +11,27 @@
+| `process.saved_user.name` | string | The username of the saved user. | `operator` |
+| `process.session_leader.pid` | int | The PID of the process's session leader. This is also the session ID (SID) of the process. | `14` |
+| `process.start` | string | The date and time the process started, in ISO 8601 format. | `2023-11-21T09:25:34.853Z` |
+| `process.user.id` | int | The effective user ID (EUID) of the process. | `1001` |
```
how this would be related to https://github.com/open-telemetry/semantic-conventions/pull/731?

trisch-me replied:
this will/might be changed later to use `user` namespace as soon as it's available and we have implemented re-use functionality.

lmolkova replied:
do I understand correctly, that the markdown won't change and we'd have `process.user.id` attribute?
Why not `user.id` - do we need more than one user?

AlexanderWert replied:
> Why not user.id - do we need more than one user?

(Assuming we come to a conclusion on https://github.com/open-telemetry/semantic-conventions/pull/731#discussion_r1511480517)
A signal could have different "users" at the same time. For example a `process.user.id` is the OS user that is running the process and in addition there could be a `client.user.id` (this is a field in ECS, not present in SemConv, yet) describing the "enduser" that authenticated through the client. So having only `user.*` without context is ambiguous.

lmolkova replied:
Leaving aside the mechanics of how inclusion/embedding/reuse can be done, what I hear that the final goal is to have `process.user.id` because `user.id` will be confusing because of all the other potential users associated with the same telemetry item. Makes sense!

mjwolf commented:
Hi @jsuereth, @lmolkova, @trisch-me, could you take a look at this PR again?  I've moved the process creation and exit fields into namespaces, added the experimental stability tag to all new fields, and resolved conflicts with main

trisch-me reviewed (approved):
thank you @mjwolf!
